cuda/multigrid/amgx_pgm_kernels.cu

```diff
@@ -111,8 +177,55 @@ template <typename ValueType, typename IndexType>
 void amgx_pgm_generate(std::shared_ptr<const CudaExecutor> exec,
                        const matrix::Csr<ValueType, IndexType> *source,
                        const Array<IndexType> &agg,
-                       matrix::Csr<ValueType, IndexType> *coarse)
-    GKO_NOT_IMPLEMENTED;
+                       matrix::Csr<ValueType, IndexType> *coarse,
+                       matrix::Csr<ValueType, IndexType> *temp)
+{
+    const auto source_nrows = source->get_size()[0];
+    const auto source_nnz = source->get_num_stored_elements();
+    const auto coarse_nrows = coarse->get_size()[0];
+    Array<IndexType> row_map(exec, source_nrows);
+    // fill coarse row pointer as zero
+    components::fill_array(exec, temp->get_row_ptrs(), coarse_nrows + 1,
+                           zero<IndexType>());
+    // compute each source row should be moved and also change column index
+    dim3 grid(ceildiv(source_nrows, default_block_size));
+    // agg source_row (for row size) coarse row source map
+    kernel::get_source_row_map_kernel<<<grid, default_block_size>>>(
+        source_nrows, agg.get_const_data(), source->get_const_row_ptrs(),
+        temp->get_row_ptrs(), row_map.get_data());
+    // prefix sum of temp_row_ptrs
+    components::prefix_sum(exec, temp->get_row_ptrs(), coarse_nrows + 1);
+    // copy source -> to coarse and change column index
+    kernel::move_row_kernel<<<grid, default_block_size>>>(
+        source_nrows, agg.get_const_data(), row_map.get_const_data(),
+        source->get_const_row_ptrs(), source->get_const_col_idxs(),
+        as_cuda_type(source->get_const_values()), temp->get_const_row_ptrs(),
+        temp->get_col_idxs(), as_cuda_type(temp->get_values()));
+    // sort csr
+    csr::sort_by_column_index(exec, temp);
```

Comment:
Maybe you can check if it is sorted, with `temp->is_sorted_by_column_index()` function of CSR, and sort it only if it not already sorted ? I guess if the `move_row_kernel` above does not give you a sorted output, then I guess you need this.

Reply:
Yes, it is usually the unsorted case.